src/test/scala/org/broadinstitute/hail/io/compress/BGzipCodecSuite.scala

```diff
@@ -1,69 +1,118 @@
+     *  - move the line boundary on two chunks by 1 character in different directions
+     *  - bgzip compressed the chunks
+     *  - stripped the empty terminate block in chunks except ad and ag (the last)
+     *  - concatenated the chunks
+     */
+    val compPath = "src/test/resources/bgz.test.sample.vcf.bgz"
 
-    val uncomp: Array[Byte] = Files.readAllBytes(Paths.get(uncompFilename))
+    val uncompHPath = new hd.fs.Path(uncompPath)
+    val compHPath = new hd.fs.Path(compPath)
 
-    val bgzipCodec = new BGzipCodec()
+    val fs = uncompHPath.getFileSystem(hadoopConf)
 
-    val conf = new Configuration()
-    val fs = FileSystem.get(URI.create(compFilename), conf)
+    val uncompIS = fs.open(uncompHPath)
+    val uncomp = IOUtils.toByteArray(uncompIS)
+    uncompIS.close()
+
+    val decompIS = new BGzipInputStream(fs.open(compHPath))
+    val decomp = IOUtils.toByteArray(decompIS)
+    decompIS.close()
 
-    val decomp = IOUtils.toByteArray(
-      bgzipCodec.createInputStream(fs.open(new Path(compFilename))))
     assert(uncomp.sameElements(decomp))
 
-    val compSize = new File(compFilename).length().toInt
-
-    for (i <- 0 until 100) {
-      val nSplits = Random.nextInt(10)
-      val splits = Array.fill[Int](nSplits)(Random.nextInt(compSize)).sorted
-      val splitRanges = (0 +: splits) zip (splits.map(_ - 1) :+ (compSize - 1))
-      // println(splitRanges.mkString(","))
-      val decomp: Array[Byte] = splitRanges.flatMap { case (start, end) =>
-        val b = new mutable.ArrayBuilder.ofByte()
-        val in = bgzipCodec.createInputStream(fs.open(new Path(compFilename)),
-          null, start, end, READ_MODE.BYBLOCK)
-        var done = false
-        do {
-          val ch = in.read()
-          if (ch != -1 && in.getPos <= end)
-            b += ch.toByte
-          else
-            done = true
-        } while (!done)
-        b.result()
-      }
-
-      assert(uncomp.sameElements(decomp))
+    val lines = Source.fromBytes(uncomp).getLines.toArray
+
+    assert(sc.textFile(uncompPath).collectOrdered()
+      .sameElements(lines))
+
+    for (i <- 1 until 20) {
+      val linesRDD = sc.textFile(compPath, i)
+      assert(linesRDD.partitions.length == i)
+      assert(linesRDD.collectOrdered().sameElements(lines))
     }
 
-    val uncompLines = sc.textFile(uncompFilename)
+    val compLength = 195353
+    val compSplits = Array[Long](6566, 20290, 33438, 41165, 56691, 70278, 77419, 92522, 106310, 112477, 112505, 124593,
+      136405, 144293, 157375, 169172, 175174, 186973, 195325)
+
+    val g = for (n <- Gen.oneOfGen(
+      Gen.choose(0, 10),
+      Gen.choose(0, 100));
+    splits <- Gen.buildableOfN[Array, Long](n,
+      Gen.oneOfGen(Gen.choose(0, compLength),
+        Gen.applyGen(Gen.oneOf[(Long) => Long](identity, _ - 1, _ + 1),
+          Gen.oneOfSeq(compSplits))))
+      .map { splits =>
+        (Array(0L, compLength) ++ splits).distinct.sorted
+      }) yield splits
```

Comment:
nitpick: I find this hard to read, though I'm not sure there's much to be done given Scala's syntax. Maybe this is a bit better?

```scala
    val g = for (n <- Gen.oneOfGen(
      Gen.choose(0, 10),
      Gen.choose(0, 100));
      validSplit = Gen.oneOfGen[Long](Gen.choose(0, compLength),
        Gen.applyGen(Gen.oneOf[(Long) => Long](identity, _ - 1, _ + 1), 
          Gen.oneOfSeq(compSplits)));
      splits <- Gen.buildableOfN[Array, Long](n, validSplit))
      yield (Array(0L, compLength) ++ splits).distinct.sorted
```